flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinator.java

```diff
@@ -0,0 +1,230 @@
+	/** The context containing the states of the coordinator. */
+	private final SourceCoordinatorContext<SplitT> context;
+	/** The split enumerator created from the associated Source. */
+	private SplitEnumerator<SplitT, EnumChkT> enumerator;
+	/** A flag marking whether the coordinator has started. */
+	private boolean started;
+
+	public SourceCoordinator(
+			ExecutorService coordinatorExecutor,
+			Source<?, SplitT, EnumChkT> source,
+			SourceCoordinatorContext<SplitT> context) {
+		this.coordinatorExecutor = coordinatorExecutor;
+		this.source = source;
+		this.enumCheckpointSerializer = source.getEnumeratorCheckpointSerializer();
+		this.splitSerializer = source.getSplitSerializer();
+		this.context = context;
+		this.enumerator = source.createEnumerator(context);
+		this.started = false;
+	}
+
+	@Override
+	public void start() throws Exception {
+		LOG.info("Starting split enumerator.");
+		enumerator.start();
+		started = true;
+	}
+
+	@Override
+	public void close() throws Exception {
+		ensureStarted();
+		LOG.info("Closing SourceCoordinator.");
+		enumerator.close();
+		coordinatorExecutor.shutdown();
+		LOG.info("Source coordinator closed.");
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public void handleEventFromOperator(int subtask, OperatorEvent event) throws Exception {
+		ensureStarted();
+		coordinatorExecutor.execute(() -> {
+			LOG.debug("Handling event from operator: {}", event);
+			if (event instanceof SourceEventWrapper) {
+				enumerator.handleSourceEvent(subtask, ((SourceEventWrapper) event).getSourceEvent());
+			} else if (event instanceof ReaderRegistrationEvent) {
+				handleReaderRegistrationEvent((ReaderRegistrationEvent) event);
+			}
+		});
+	}
+
+	@Override
+	public void subtaskFailed(int subtaskId) {
+		ensureStarted();
+		coordinatorExecutor.execute(() -> {
+			LOG.info("Handling subtask {} failure.", subtaskId);
+			List<SplitT> splitsToAddBack = context.getAndRemoveUncheckpointedAssignment(subtaskId);
+			context.unregisterSourceReader(subtaskId);
+			LOG.debug("Adding {} back to the split enumerator.", splitsToAddBack);
+			enumerator.addSplitsBack(splitsToAddBack, subtaskId);
+		});
+	}
+
+	@Override
+	public CompletableFuture<byte[]> checkpointCoordinator(long checkpointId) throws Exception {
+		ensureStarted();
+		return CompletableFuture.supplyAsync(() -> {
+			try {
+				LOG.debug("Taking a state snapshot for checkpoint {}", checkpointId);
+				return toBytes(checkpointId);
+			} catch (Exception e) {
+				throw new FlinkRuntimeException("Failed to checkpoint coordinator due to ", e);
+			}
+		}, coordinatorExecutor);
+	}
+
+	@Override
+	public void checkpointComplete(long checkpointId) {
+		ensureStarted();
+		coordinatorExecutor.execute(() -> {
+			LOG.info("Marking checkpoint {} as completed.", checkpointId);
+			context.onCheckpointComplete(checkpointId);
+		});
+	}
+
+	@Override
+	public void resetToCheckpoint(byte[] checkpointData) throws Exception {
+		if (started) {
```

Comment:
Do we want to enforce this? We may want to reset to checkpoint whenever the scheduler decides to do a global failover.

Global failovers are the scheduler's "safety net". When during common failover (pipelined region) something unexpected happens (bug, inconsistency, or so) then it does a global failover (reset everything to latest checkpoint). That has a high chance of fixing inconsistencies. It is costly, but better than becoming unavailable.

When such a situation happens, I think we should reset the enumerator, because the `subtaskFailed()`-based bookkeeping may be inconsistent as well.